### PR TITLE
HaCreator - 64 Bit Client

### PR DESCRIPTION
**Pull request by SoulGirlJP**
I made an option to switch between 32 bit client and 64 bit client compatible, maybe its not optimal way and could just make the thing detect if it contains a Data folder, but anyways here is my take feel free to make it better, there is couple of issues like map simulator and some tile display issue not working, but i think the rest should be fine, it might need to get looked into it more since obviously not that perfect, like Map9_001.wz still needs to be loaded as i didn`t find a way to do this couple of things missing, but wanted to contribute to the community much love <3

### CODE REVIEW

**Comment by lastbattle** [comment]
Hi,  I would appreciate if you are able to pull from the latest master branch, then opening up another pull request. 
It is overriding some of my prior changes.

Thanks a lot!

**Comment by SoulGirlJP** [comment]
Ohhh yeh sure

**Comment by SoulGirlJP** [comment]
Yeh i dunno why i am having such a hard time with Github

**Comment by lastbattle** [comment]
It might be easier by just pulling the latest branch, and manually making the changes again 🤔

**Comment by SoulGirlJP** [comment]
Hmm yeh ill just do it next week

**Comment by SoulGirlJP** [comment]
After a long break i have decided to come back and pulled the latest changes as asked before, and i completely redid since i had a lot of issues with the last pull so i just pulled and redid all the code necessary to make it work again, as mentioned there is a lot of work to do still, i just wanted to help out the community and help the open source as best to my ability, if there is any questions please let me know! Much love <3

**Comment by lastbattle** [comment]
I will take a look this weekend! Thanks a lot for your help, I appreciate it!!

File: HaCreator/GUI/Initialization.cs [hunk 1]
```diff
@@ -122,63 +150,189 @@ private void InitializeWzFiles(string wzPath, WzMapleVersion fileVersion)
...
+                    }
+                } else
+                {
+                    foreach (string soundWzFile in WzFileManager.SOUND_WZ_FILES)
+                    {
+                        textBox2.Text = string.Format("Initializing {0}.wz...", soundWzFile);
+                        Application.DoEvents();
+                        Program.WzManager.LoadWzFile(soundWzFile.ToLower());
+                        Program.WzManager.ExtractSoundFile(soundWzFile.ToLower());
+                    }
```
Comment (shailist) [inline HaCreator/GUI/Initialization.cs:240]: the only difference is which files we iterate, the loop is the same:
```cs
// Load sound
var soundWzFiles = isClient64() ? WzFileManager.SOUND_WZ_FILES_64 : WzFileManager.SOUND_WZ_FILES;
foreach (string soundWzFile in soundWzFiles )
{
    textBox2.Text = string.Format("Initializing {0}.wz...", soundWzFile);
    Application.DoEvents();
    Program.WzManager.LoadWzFile(soundWzFile.ToLower());
    Program.WzManager.ExtractSoundFile(soundWzFile.ToLower());
}
```

File: HaCreator/GUI/Initialization.cs [hunk 2]
```diff
@@ -122,63 +150,189 @@ private void InitializeWzFiles(string wzPath, WzMapleVersion fileVersion)
...
+                    {
+                        if (Program.WzManager.LoadWzFile(mapwzFile.ToLower()))
+                        {
+                            textBox2.Text = string.Format("Initializing {0}.wz...", mapwzFile);
+                            Application.DoEvents();
+                            Program.WzManager.ExtractBackgroundSets();
+                            Program.WzManager.ExtractObjSets();
+                        }
+                    }
+                }
```
Comment (shailist) [inline HaCreator/GUI/Initialization.cs:320]:
```cs
var mapWzFiles = isClient64() ? WzFileManager.MAP_WZ_FILES_64 : WzFileManager.MAP_WZ_FILES;
foreach (string mapwzFile in mapWzFiles )
{
    if (Program.WzManager.LoadWzFile(mapwzFile.ToLower()))
    {
        textBox2.Text = string.Format("Initializing {0}.wz...", mapwzFile);
        Application.DoEvents();
        Program.WzManager.ExtractBackgroundSets();
        Program.WzManager.ExtractObjSets();
    }
}
```

File: HaCreator/MapEditor/Info/MobInfo.cs [hunk 3]
```diff
@@ -57,25 +58,49 @@ public override void ParseImage()
...
+                        mobImage.HCTag = MobInfo.Load(mobImage);
+                    }
+                    MobInfo result = (MobInfo)mobImage.HCTag;
+                    result.ParseImageIfNeeded();
+                    return result;
                 }
-                MobInfo result = (MobInfo)mobImage.HCTag;
-                result.ParseImageIfNeeded();
-                return result;
+                return null;
```
Comment (shailist) [inline HaCreator/MapEditor/Info/MobInfo.cs:102]:
```cs
var mobWzFiles = Initialization.isClient64() ? WzFileManager.MOB_WZ_FILES_64 : WzFileManager.MOB_WZ_FILES;
foreach (string mobWzFile in mobWzFiles)
{
    WzImage mobImage = (WzImage)Program.WzManager[mobWzFile.ToLower()]?[id + ".img"];
    if (mobImage == null)
        continue;

    if (!mobImage.Parsed)
    {
        mobImage.ParseImage();
    }
    if (mobImage.HCTag == null)
    {
        mobImage.HCTag = MobInfo.Load(mobImage);
    }
    MobInfo result = (MobInfo)mobImage.HCTag;
    result.ParseImageIfNeeded();
    return result;
}
return null;
```

**Review by shailist — CHANGES_REQUESTED** [review]
glossed around the changes and left some suggestions for cleaner code.
didn't really look into the internals and didn't verify if it even worked :shrug:, but looks good

**Comment by Alon-Sachs** [comment]
is there a way to add this to harepacker aswell?

**Comment by SoulGirlJP** [comment]
Harepacker already works

**Comment by Lingko621** [comment]
how can i use this program? and it works?
if it works  send me file please  ToT

**Comment by lastbattle** [comment]
Looking at it now! Was away.

**Comment by lastbattle** [comment]
Merged! I'll do a little clean up later, once again thanks for the contribution!

**Comment by SoulGirlJP** [comment]
You are very welcome! I'm happy i was able to help!

File: HaCreator/MapEditor/Info/MobInfo.cs (hunk 3 above)
Comment (lastbattle) [reply HaCreator/MapEditor/Info/MobInfo.cs:102]: Pushed! thanks.
https://github.com/lastbattle/Harepacker-resurrected/commit/b37108aca618bfdf705b785f5090e6c5ecae7901

**Comment by lastbattle** [comment]
I've made it better, feel free to take a look! 
It'll allow for better support for future versions of .wz files, as the location of the images, and directories are not static.

https://github.com/lastbattle/Harepacker-resurrected/commit/cee77e2c286ce77966cfe93118cc54004acc93f1

**Comment by SoulGirlJP** [comment]
Amazing work, i wanted to do this too but i felt lazy haha good job!

**Comment by lastbattle** [comment]
Exactly!! It was just so tedious.
That was the reason why I avoided doing it months after 64-client was released.
It kept being pushed back and at the same time I was busy on other personal stuff. I guess this pull request kind of forced me to 😀

**Comment by SoulGirlJP** [comment]
Ahhh i see yeah well it's understandable hahahaha

**Pull request merged**